mutiny-core/src/blindauth.rs

```diff
@@ -0,0 +1,332 @@
+    pub service_id: u32,
+    pub name: String,
+    pub allocation_count: u32,
+    pub allocation_type: String,
+    pub subscription_plan_reference: Option<i32>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RedeemServiceTokenRequest {
+    pub service_id: u32,
+    pub plan_id: u32,
+    pub blinded_message: BlindedMessage,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RedeemServiceTokenResponse {
+    pub service_id: u32,
+    pub plan_id: u32,
+    pub blind_sig: BlindedSignature,
+}
+
+pub struct BlindAuthClient<S: MutinyStorage> {
+    secret: DerivableSecret,
+    auth_client: Arc<MutinyAuthClient>,
+    base_url: String,
+    storage: S,
+    token_storage: Arc<RwLock<TokenStorage>>,
+    pub logger: Arc<MutinyLogger>,
+}
+
+impl<S: MutinyStorage> BlindAuthClient<S> {
+    pub fn new(
+        xprivkey: ExtendedPrivKey,
+        auth_client: Arc<MutinyAuthClient>,
+        network: Network,
+        base_url: String,
+        storage: &S,
+        logger: Arc<MutinyLogger>,
+    ) -> Result<Self, MutinyError> {
+        let token_storage = storage.get_token_storage()?;
+        let secret = create_blind_auth_secret(xprivkey, network)?;
+
+        Ok(Self {
+            secret,
+            auth_client,
+            base_url,
+            storage: storage.clone(),
+            token_storage: Arc::new(RwLock::new(token_storage)),
+            logger,
+        })
+    }
+
+    pub async fn redeem_available_tokens(&self) -> Result<(), MutinyError> {
+        // check to see what is available to the user
+        let available_tokens = self.check_available_tokens().await?;
+
+        // fetch available one by one
+        for service in available_tokens.tokens {
+            match self.redeem_token(service).await {
+                Ok(_) => (),
+                Err(e) => {
+                    log_error!(self.logger, "could not redeem token: {e}");
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    async fn check_available_tokens(&self) -> Result<CheckServiceTokenResponse, MutinyError> {
+        get_available_tokens(&self.auth_client, &self.base_url).await
+    }
+
+    async fn redeem_token(&self, service: ServicePlans) -> Result<BlindedToken, MutinyError> {
+        let service_id = service.service.id;
+        let plan_id = service.plan.id;
+        let mut token_storage_guard = self.token_storage.write().await;
+        let next_counter = token_storage_guard.get_value(service_id, plan_id) + 1;
+
+        // create the deterministic info to derive the token from
+        let token_to_blind =
+            derive_blind_token(&self.secret, service_id, plan_id, next_counter).await?;
+        let token_req = RedeemServiceTokenRequest {
+            service_id,
+            plan_id,
+            blinded_message: token_to_blind.blinded_message,
+        };
+
+        // request a blinded signature
+        let token_resp = post_redeem_tokens(&self.auth_client, &self.base_url, token_req).await?;
```

Comment:
we shouldn't use our `auth_client` here, otherwise the request will have the JWT token and all the blind token stuff will be for nothing

Reply:
We have to here. This is to get a blind signature from the server. In order to know if we can issue a blind signature, we need to know if the user has not been given one before.

Perhaps the name is confusing? The actual redemption of the service token isn't done here but will be done with the individual service clients. In the hermes client, the unblinding part will take place and that's what needs to be done without the auth client.

Reply:
Ah yeah I see, the naming confused me thinking you are redeeming the tokens but you are really just getting them signed.

Reply:
gotcha, yeah i'll fix the names